File: src/parser/parser.cpp
#include "parser/node.h"
#include "parser/parser.h"
#include "parser/statements/file-statement.h"
#include "fault.h"

Parser::Parser(vector<Token>& tokens) 
: tokens(tokens), i(0), tree(nullptr), context() {}

Parser::~Parser() {
    delete tree;
}

str Parser::toString() const {
    return tree? tree->toString() : "";
}

void Parser::makeTree() {
    FileStatement* node = new FileStatement();
    node->makeTree();
    tree = node;
}

Token& Parser::next() const {
    return tokens[(i > tokens.size() - 1)? tokens.size() - 1: i];
}

Token& Parser::ahead(nat pos) const {
    nat i = self.i + pos;
    return tokens[(i > tokens.size() - 1)? tokens.size() - 1: i];
}

Token& Parser::take() {
    Token& token = next();
    i += 1;
    return token;
}

bool Parser::nextOrAfterHas(vector<str> values) {
    return next().has(values) or ahead(1).has(values);
}

Token& Parser::expectingOf(vector<Token::Type> types) {
    if(next().of(types)) return take();

    vector<str> names;
    for(Token::Type type : types) names.push_back(Token::typeToString(type));

    str message = fmt::format("Expecting one of [{}]", join(names, ", "));
    throw Fault::fail(take(), message);
}

Token& Parser::expectingHas(vector<str> values) {
    if(next().has(values)) return take();

    vector<str> escaped;
    for(str value : values) escaped.push_back(fmt::format("\"{}\"", value));

    str message = fmt::format("Expecting one of [{}]", join(escaped, ", "));
    throw Fault::fail(take(), message);
}

void Parser::skipNewlines() {
    while(next().has({"\n"})) take();
}

File: app/src/main/kotlin/sea/parser/Parser.kt
package sea.parser

import sea.Faults
import sea.lexer.Token
import sea.grammar.FileStatement

data class Parser(val faults: Faults, val tokens: ArrayList<Token>) {
    var i = 0
    var tree: Node = FileStatement()

    override fun toString() = "$tree"

    fun makeTree() {
        (tree as FileStatement).makeTree(this)
    }

    val next: Token
        get() = tokens[if(i > tokens.size - 1) tokens.size - 1 else i]

    fun ahead(pos: Int): Token {
        val i = this.i + pos
        return tokens[if(i > tokens.size - 1) tokens.size - 1 else i]
    }

    fun take(): Token {
        val token = next
        i += 1
        return token
    }

    fun expectingOf(vararg types: TokenType): Token {
        if(next.of(*types)) return take()

        val names = (types.map { it.name }).joinToString(", ", "[", "]")
        throw faults.fail(take(), "Expecting type of $names")
    }

    fun expectingHas(vararg values: String): Token {
        if(next.has(*values)) return take()

        val vals = values.joinToString(", ", "[", "]")
        throw faults.fail(take(), "Expecting value of $vals".replace("\n", "\\n"))
    }

    fun expectingOf(c: Collection<TokenType>) = expectingOf(*c.toTypedArray())
    fun expectingHas(c: Collection<String>) = expectingHas(*c.toTypedArray())

    fun nextOrAfterHas(vararg values: String): Boolean {
        return next.has(*values) or ahead(1).has(*values)
    }

    fun nextOrAfterHas(c: Collection<String>) = nextOrAfterHas(*c.toTypedArray())

    fun skipNewlines() {
        while(next.has("\n")) take()
    }

    companion object {
        val lineEnds = Token.LINE_ENDS.toTypedArray()
    }
}
